pkg/client/apiutil/restmapper.go

```diff
@@ -194,13 +209,28 @@ func (m *mapper) addKnownGroupAndReload(groupName string, versions ...string) er
 		return fmt.Errorf("failed to get API group resources: %w", err)
 	}
 
-	if _, ok := m.knownGroups[groupName]; ok {
-		groupResources = m.knownGroups[groupName]
-	}
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.addGroupVersionResourcesToCache(groupVersionResources)
+	return nil
+}
 
+// addGroupVersionResourcesToCache does what the name suggests. The mutex must be held when
+// calling it.
+func (m *mapper) addGroupVersionResourcesToCache(gvr map[schema.GroupVersion]*metav1.APIResourceList) {
 	// Update information for group resources about the API group by adding new versions.
-	// Ignore the versions that are already registered.
-	for groupVersion, resources := range groupVersionResources {
+	// Ignore the versions that are already registered
```

Comment:
Should we return early if len(groupVersionResources) == 0? (I think in that case it would recreate the mapper for nothing, basically a no-op)

Reply:
What codepath do you see where this could happen (len(gvk) == 0)?

Reply:
I was thinking about
```go
m.addGroupVersionResourcesToCache(maybeResources)
```
for the case where aggregated discovery is not enabled

Reply:
Added a new comment with a different idea